### ISSUE TITLE
Running js-ipfs behind a proxy

### ISSUE THREAD

**Issue by JGAntunes** [ipfs/js-ipfs#1865]
Hey there! :wave:

**tl;dr;** Is there a way for a node to "advertise" one swarm address without binding to it?

I've been looking into running some tests with multiple [IPFS nodes running behind a TCP proxy](https://github.com/JGAntunes/helm-charts/tree/master/ipfs-testbed). I'm using a bootstrap discovery method with 3 initial nodes deployed followed by another set of them. The config I'm using is something like this:
```
{
    "Addresses": {
        "Swarm": [
            "/ip4/127.0.0.1/tcp/4002"
        ],
        "API": "/ip4/172.17.0.5/tcp/5002",
(...)
```
And the TCP proxy running at port `8002` forwarding traffic to the ipfs swarm address.

I came to find however that given the peer doesn't know its own address, other peers end up with peer lists like this (notice the two peers with a  `127.0.0.1` address):
```
kubectl exec -it node-3-ipfs-testbed-bddb7b794-rftkz --container=js-ipfs -- jsipfs swarm peers
/ip4/127.0.0.1/tcp/4002/ipfs/QmSFkkfUZU3zZX3uEXm8FxJMpMGnU1qejRQUiy6UzyDJWk
/ip4/127.0.0.1/tcp/4002/ipfs/QmfQUpk41XPjQ7AehCSbRi7CCRB8LERB1pNxQFHMk9JiUY
/dns4/node-2-ipfs-testbed/tcp/8002/ipfs/QmcTaFLT2vziRVSN8nnd3HRhZxXbo8MG1ghRc65qHCkZ8v
/dns4/node-1-ipfs-testbed/tcp/8002/ipfs/QmaZfXvwHvNk3nrJyrgPWDi1LK6bPDPr5JSXXWV88Hykts
```

If I try to use the proxy address in the swarm config it will fail because the node tries to bind to it. So I guess what I need is some way for the peer to list an address without binding to it, is this possible? I was looking around in `libp2p` for something that would  allow this behaviour but to no luck :disappointed: do any of you know a way around this?

**Comment by alanshaw** [comment]
Ha, great question. Can we use a circuit address for this @jacobheun / @vasco-santos ?

**Comment by JGAntunes** [comment]
Meanwhile kept searching and found this - https://github.com/multiformats/js-multiaddr/blob/master/src/index.js#L182 - I wonder if this is a case were address encapsulation would be useful? :thinking:

**Comment by jacobheun** [comment]
This is where https://github.com/libp2p/js-libp2p/issues/202 would be very useful. It would provide the ability to set the listen/announce/no announce addresses without messing with the addresses after startup.

@JGAntunes encapsulation/decapsulation is likely the best workaround without adding the config changes to libp2p, which would be the best option.

The encapsulation workaround will require some post processing on the peerInfo multiaddrs after libp2p has started. We currently do something similar on boot for the transport listeners in libp2p-switch, https://github.com/libp2p/js-libp2p-switch/blob/v0.41.5/src/transport.js#L171.  You could do the replace with the encapsulated address, which would be your proxy address `/dns4/node-2-ipfs-testbed/tcp/8002` and the ipfs address of the node `/ipfs/Qm`. This should also allow you to prevent advertising the local address. An issue you might hit here is advertising addresses before you've updated them.

**Comment by JGAntunes** [comment]
@jacobheun yes that would be perfect! Thanks for pointing that out :+1: I don't have much free time on my hands but maybe I can help out with that (moving this discussion to the issue).

Meanwhile I'll probably take your advice with the second approach so that I can get quickly unblocked. Thanks for all the pointers again :pray:

**Comment by JGAntunes** [comment]
And I think it's ok to close this :+1: thank you @jacobheun and @alanshaw.
